File: backend/internal/models/product.go
package models

import (
	"time"

	"go.mongodb.org/mongo-driver/bson/primitive"
)

type Review struct {
	ID        primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	Name      string             `json:"name" bson:"name"`
	Rating    float64            `json:"rating" bson:"rating"`
	Comment   string             `json:"comment" bson:"comment"`
	User      primitive.ObjectID `json:"user" bson:"user"`
	CreatedAt time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt time.Time          `json:"updatedAt" bson:"updatedAt"`
}

type Product struct {
	ID           primitive.ObjectID `json:"_id,omitempty" bson:"_id,omitempty"`
	Name         string             `json:"name" bson:"name"`
	Image        string             `json:"image" bson:"image"`
	Brand        string             `json:"brand" bson:"brand"`
	Quantity     int                `json:"quantity" bson:"quantity"`
	Category     primitive.ObjectID `json:"category" bson:"category"`
	Description  string             `json:"description" bson:"description"`
	Reviews      []Review           `json:"reviews" bson:"reviews"`
	Rating       float64            `json:"rating" bson:"rating"`
	NumReviews   int                `json:"numReviews" bson:"numReviews"`
	Price        float64            `json:"price" bson:"price"`
	CountInStock int                `json:"countInStock" bson:"countInStock"`
	CreatedAt    time.Time          `json:"createdAt" bson:"createdAt"`
	UpdatedAt    time.Time          `json:"updatedAt" bson:"updatedAt"`
}

File: frontend/app/schemas/product.ts
import { z } from "zod";
import { reviewSchema } from "./review";

export const productSchema = z.object({
  _id: z.string().optional(),
  name: z.string().min(1),
  image: z.string().min(1),
  brand: z.string().min(1),
  quantity: z.number().int().min(0),
  categoryId: z.string(), // ObjectId as string
  description: z.string().min(1),
  reviews: z.array(reviewSchema).default([]),
  rating: z.number().min(0).max(5).default(0),
  numReviews: z.number().int().min(0).default(0),
  price: z.number().min(0).default(0),
  countInStock: z.number().int().min(0).default(0),
  createdAt: z.iso.datetime().optional(),
  updatedAt: z.iso.datetime().optional(),
});

export type Product = z.infer<typeof productSchema>;

export const newProduct = (): Product => ({
  name: "",
  image: "",
  brand: "",
  quantity: 0,
  categoryId: "",
  description: "",
  reviews: [],
  rating: 0,
  numReviews: 0,
  price: 0,
  countInStock: 0,
  createdAt: "",
  updatedAt: "",
});

export const getId = (item: Product) => item._id || "";

export type ProductFilters = {
  category?: string;
  price?: string;
  brand?: string;
};
